### ISSUE TITLE
step1 models are always the same as step0 models

### ISSUE THREAD

**Issue by StellaAthena** [EleutherAI/gpt-neox#1373]
When using warmup, the lr is set to 0 for the very first iteration. This means that the models are unchanged between step 0 and step 1. You can see this in every Pythia model on the HF Hub, for example.

This happens because in `return float(self.start_lr) * num_iters_ / self.warmup_iter` [here](https://github.com/EleutherAI/gpt-neox/blob/d12c771198388980ee054617e537665f044e0584/megatron/learning_rates.py#L70) `num_iters_` is 0 since the first step hasn't finished yet. The same happens in Megatron-DeepSpeed.

Megatron-LM uses `return self.init_lr + (max_lr - self.init_lr) * float(self.num_steps) / float(self.lr_warmup_steps))` (see [this line](https://github.com/NVIDIA/Megatron-LM/blob/4578ed8a6f66a1fe64e81c7b52dd490f6fe6b33a/megatron/core/optimizer_param_scheduler.py#L228)). The `init_lr` term could be used to prevent this, but its very difficult to use (you don't set it to 1, you set it to the actual lr you want at step 1) and it defaults to 0 anyways.

If this is standard then it's not a bug *per se*, but it surprised me. At a mimum it should be documented behavior.
